mri/reconstructors/self_calibrating.py

```diff
@@ -0,0 +1,180 @@
+    mu: float or np.ndarray, (optional, default 0)
+        If prox_op is None, the value of mu is used to form a proximity
+        operator that is soft thresholding of the wavelet coefficients.
+        If prox_op is specified, this is ignored.
+    lips_calc_max_iter: int, default 10
+        Defines the maximum number of iterations to calculate the lipchitz
+        constant
+    num_check_lips: int, default 10
+        Number of iterations to check if the lipchitz constant is correct
+    lipschitz_cst: int, default None
+        The user specified lipschitz constant. If this is not specified,
+        it is calculated using PowerMethod
+    kspace_portion: int or tuple (default is 0.1 in all dimension)
+        int or tuple indicating the k-space portion used to estimate the coil
+        sensitivity information.
+        if int, will be evaluated to (0.1,)*nb_dim of the image
+    smaps_extraction_mode: string 'FFT' | 'NFFT' | 'Stack' | 'gridding' default
+        Defines the mode in which we would want to interpolate to extract the
+        sensitivity information,
+        NOTE: FFT should be considered only if the input has
+        been sampled on the grid
+    smaps_gridding_method: string 'linear' (default) | 'cubic' | 'nearest'
+        For gridding mode, it defines the way interpolation must be done used
+        by the sensitivity extraction method.
+    n_jobs: int, default 1
+        The number of CPUs used to accelerate the reconstruction
+    verbose: int, default 0
+        Verbosity level.
+            1 => Print basic debug information
+            5 => Print all initialization information
+            20 => Calculate cost at the end of each iteration.
+                NOTE : This is computationally intensive.
+            30 => Print the debug information of operators if defined by class
+    Note:
+    -----
+    The user is expected to specify the either prox_op or mu to obtain
+    reconstructions, else the above equations lose the regularization terms
+    resulting in inverse transform as solution.
+    The reconstruction in this case proceeds with a warning.
+    """
+
+    def __init__(self, fourier_op, linear_op=None, prox_op=None, mu=0,
+                 gradient_method="synthesis", lips_calc_max_iter=10,
+                 num_check_lips=10, lipschitz_cst=None, kspace_portion=0.1,
+                 smaps_extraction_mode='gridding',
+                 smaps_gridding_method='linear', n_jobs=1, verbose=0):
+        if linear_op is None:
+            linear_op = WaveletN(
+                wavelet_name="sym8",
+                nb_scale=3,
+                dim=len(fourier_op.shape),
+                n_coils=1,
+                verbose=bool(verbose >= 30),
+            )
+        # Ensure that we are in right multichannel config
+        if linear_op.n_coils != 1:
+            raise ValueError("The value of n_coils for linear operation must "
+                             "be 1 for Self-Calibrating reconstruction!")
```

Comment:
Could you add on the error message something like : `For single channel reconstruction please consider ReconstructorBase` ?

Reply:
That doesnt make sense, this error comes up if the user specified n_coils is not 1, but he uses self calibrating reconstruction. If needed we can redirect towards calibrationless, but that may be totally unknown.

Reply:
Yes you are partly right, the error message doesn't make any sense however, [some reconstruction method](https://ieeexplore.ieee.org/abstract/document/7229332) uses self-calibrating approaches with other norms such as the group-LASSO. this is way specifying the L1-norm isn't a good idea

Reply:
Any other ideas for the error message that is more general? Also, in this case, our check would get more complicated and not mere check on n_coils being 1. 

We do agree that the first error message given above is not right also right?

Reply:
> Any other ideas for the error message that is more general? Also, in this case, our check would get more complicated and not mere check on n_coils being 1.

Just checking that its either 1 or fourier_op.n_coils doesn't make it more complicated.
> We do agree that the first error message given above is not right also right?
Yes

For the error message probably something more like: Number of coils in the linear operator is not aligned with the number of coils in the Fourier Operator

Reply:
> Number of coils in the linear operator is not aligned with the number of coils in the Fourier Operator

But that also is not right, that would make user think the the n_coils for linear and fourier operator must be the same. But the check here is that for this particular case of self calibrating reconstruction, the n_coils must be unity (for proximity defined by sparse threshold)